### PR TITLE
enh(resources): handle categories resource detail panel

### PR DESCRIPTION
**Pull request by adr-mo**
## Description

This PR intends to add a new entry to detail endpoints for host/service resources - `categories`

**Fixes** # (issue)

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 20.10.x
- [ ] 21.04.x
- [ ] 21.10.x
- [x] 22.04.x (master)

<h2> How this pull request can be tested ? </h2>

See jira details

## Checklist

- [x] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

**Comment by adr-mo** [comment]
@adr-mo Rewrite unit tests in pest syntax

File: src/Core/Infrastructure/RealTime/Api/Hypermedia/MetaServiceHypermediaProvider.php [hunk 1]
```diff
@@ -201,6 +201,14 @@ public function createInternalGroupsUri(mixed $response): array
         return [];
     }
 
+    /**
+     * @inheritDoc
+     */
+    public function createInternalCategoriesUri(mixed $response): array
+    {
+        return [];
```
Comment (jeremyjaouen) [inline src/Core/Infrastructure/RealTime/Api/Hypermedia/MetaServiceHypermediaProvider.php:209]: Is this empty array normal ?

File: src/Core/Infrastructure/RealTime/Repository/Tag/DbReadTagRepository.php [hunk 2]
```diff
@@ -22,33 +22,71 @@
...
 
     /**
      * @inheritDoc
      */
     public function findAllByTypeId(int $typeId): array
     {
-        $this->info('Fetching tags from database');
+        $this->info('Fetching tags from database of type', ['type' => $typeId]);
+
+        $request = 'SELECT id, name, `type`
```
Comment (kduret) [inline src/Core/Infrastructure/RealTime/Repository/Tag/DbReadTagRepository.php:63]: add sql calc found rows

File: tests/php/Core/Application/RealTime/UseCase/FindService/FindServiceTest.php [hunk 3]
```diff
@@ -22,495 +22,386 @@
...
+    $this->hostRepository
+        ->expects($this->once())
+        ->method('findHostById')
+        ->willReturn(null);
+
+    $findServicePresenter = new FindServicePresenter($this->hypermediaCreator, $this->presenterFormatter);
+
+    $findService(1, 20, $findServicePresenter);
+
+    $this->assertEquals($findServicePresenter->getResponseStatus(), new NotFoundResponse('Host'));
```
Comment (kduret) [inline tests/php/Core/Application/RealTime/UseCase/FindService/FindServiceTest.php:102]: use èxpect

File: src/Centreon/Infrastructure/Broker/BrokerRepositoryRDB.php [hunk 4]
```diff
@@ -53,52 +53,53 @@ public function __construct(DatabaseConnection $db)
...
         $statement->execute();
-
         $brokerConfigurations = [];
         while (($result = $statement->fetch(\PDO::FETCH_ASSOC)) !== false) {
             $brokerConfigurations[] = (new BrokerConfiguration())
-                ->setId((int) $result['id'])
-                ->setConfigurationKey($parameterName)
-                ->setConfigurationValue($result['config_value']);
+                ->setId($result['config_id'])
+                ->setConfigurationKey('bbdo_version')
```
Comment (kduret) [inline src/Centreon/Infrastructure/Broker/BrokerRepositoryRDB.php:101]: use `BBDO_VERSION_CONFIG_KEY` constant

File: src/Core/Application/Configuration/Broker/BrokerBBDO.php [hunk 5]
```diff
@@ -43,13 +44,14 @@ public function __construct(protected BrokerRepositoryInterface $brokerRepositor
...
+        $brokerConfigurations = $this->brokerRepository->findAllConfigurations();
         foreach ($brokerConfigurations as $brokerConfiguration) {
             if (
-                version_compare(
+                VersionHelper::compare(
                     (string) $brokerConfiguration->getConfigurationValue(),
-                    self::MINIMUM_BBDO_VERSION_SUPPORTED
-                ) > 0
+                    self::MINIMUM_BBDO_VERSION_SUPPORTED,
+                    VersionHelper::GT
```
Comment (kduret) [inline src/Core/Application/Configuration/Broker/BrokerBBDO.php:53]:
```suggestion
                    VersionHelper::GE
```

File: src/Core/Application/RealTime/UseCase/FindHost/FindHost.php [hunk 6]
```diff
@@ -98,8 +104,12 @@ public function __invoke(int $hostId, FindHostPresenterInterface $presenter): vo
...
         }
 
-        foreach ($hostgroups as $hostgroup) {
-            $host->addHostgroup($hostgroup);
+        $host->setHostgroups($hostgroups);
+
+        $categories = $this->tagRepository->findAllByResourceAndTypeId($host->getId(), 0, Tag::HOST_CATEGORY_TYPE_ID);
+
+        foreach ($categories as $category) {
+            $host->addCategory($category);
```
Comment (kduret) [inline src/Core/Application/RealTime/UseCase/FindHost/FindHost.php:112]: use setCategories

File: src/Core/Application/RealTime/UseCase/FindHost/FindHost.php [hunk 7]
```diff
@@ -98,8 +104,12 @@ public function __invoke(int $hostId, FindHostPresenterInterface $presenter): vo
             $hostgroups = $this->hostgroupRepository->findAllByHostIdAndAccessGroupIds($hostId, $accessGroupIds);
         }
 
-        foreach ($hostgroups as $hostgroup) {
-            $host->addHostgroup($hostgroup);
+        $host->setHostgroups($hostgroups);
```
Comment (kduret) [inline src/Core/Application/RealTime/UseCase/FindHost/FindHost.php:107]:
```suggestion
        $host->setGroups($hostgroups);
```

File: src/Core/Domain/RealTime/Model/Host.php [hunk 8]
```diff
@@ -681,4 +687,47 @@ public function getCheckAttempts(): ?int
...
+     */
+    public function addCategory(Tag $category): self
+    {
+        $this->categories[] = $category;
+        return $this;
+    }
+
+    /**
+     * @param Tag[] $categories
+     * @return self
```
Comment (kduret) [inline src/Core/Domain/RealTime/Model/Host.php:711]: add `throws`

File: src/Core/Domain/RealTime/Model/Host.php [hunk 9]
```diff
@@ -681,4 +687,47 @@ public function getCheckAttempts(): ?int
...
+     */
+    public function setCategories(array $categories): self
+    {
+        foreach ($categories as $category) {
```
Comment (kduret) [inline src/Core/Domain/RealTime/Model/Host.php:715]:
```suggestion
        $this->categories = [];
        foreach ($categories as $category) {
```

File: src/Core/Domain/RealTime/Model/Service.php [hunk 10]
```diff
@@ -633,4 +638,48 @@ public function setHasGraphData(bool $hasGraphData): self
...
+        return $this;
+    }
+
+    /**
+     * @param Tag[] $categories
+     * @return self
+     */
+    public function setCategories(array $categories): self
+    {
+        foreach ($categories as $category) {
```
Comment (kduret) [inline src/Core/Domain/RealTime/Model/Service.php:666]:
```suggestion
    {
        $this->categories = [];
        foreach ($categories as $category) {
```

File: src/Core/Domain/RealTime/Model/Host.php [hunk 11]
```diff
@@ -681,4 +687,47 @@ public function getCheckAttempts(): ?int
...
+        return $this;
+    }
+
+    /**
+     * @param Hostgroup[] $hostgroups
+     * @return self
+     */
+    public function setHostgroups(array $hostgroups): self
+    {
+        foreach ($hostgroups as $hostgroup) {
```
Comment (kduret) [inline src/Core/Domain/RealTime/Model/Host.php:727]:
```suggestion
    {
        $this->groups = [];
        foreach ($hostgroups as $hostgroup) {
```

File: src/Core/Domain/RealTime/Model/Service.php [hunk 12]
```diff
@@ -633,4 +638,48 @@ public function setHasGraphData(bool $hasGraphData): self
...
+        }
+
+        return $this;
+    }
+
+    /**
+     * @param Servicegroup[] $servicegroups
+     * @return self
+     */
+    public function setServicegroups(array $servicegroups): self
```
Comment (kduret) [inline src/Core/Domain/RealTime/Model/Service.php:677]:
```suggestion
    public function setGroups(array $servicegroups): self
```

File: src/Core/Domain/RealTime/Model/Service.php [hunk 13]
```diff
@@ -633,4 +638,48 @@ public function setHasGraphData(bool $hasGraphData): self
...
+    {
+        foreach ($servicegroups as $servicegroup) {
```
Comment (kduret) [inline src/Core/Domain/RealTime/Model/Service.php:679]:
```suggestion
    {
        $this->groups = [];
        foreach ($servicegroups as $servicegroup) {
```

File: src/Core/Infrastructure/RealTime/Api/Hypermedia/HypermediaCreatorHelper.php [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    /**
+     * Checks if contact has access to pages defined in roles
+     *
+     * @param ContactInterface $contact
+     * @param string[] $topologyRoles
+     * @return boolean
+     */
+    public function canContactAccessPages(ContactInterface $contact, array $topologyRoles): bool
```
Comment (kduret) [inline src/Core/Infrastructure/RealTime/Api/Hypermedia/HypermediaCreatorHelper.php:49]: move into an other class which is not related to hypermedia

Comment (kduret) [reply src/Core/Infrastructure/RealTime/Api/Hypermedia/HypermediaCreatorHelper.php:49]: security scope

File: src/Core/Infrastructure/RealTime/Api/Hypermedia/HypermediaCreatorHelper.php [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+declare(strict_types=1);
+
+namespace Core\Infrastructure\RealTime\Api\Hypermedia;
+
+use Centreon\Domain\Contact\Contact;
+use Centreon\Domain\Contact\Interfaces\ContactInterface;
+use Core\Infrastructure\Common\Api\HttpUrlTrait;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+
+class HypermediaCreatorHelper
```
Comment (kduret) [inline src/Core/Infrastructure/RealTime/Api/Hypermedia/HypermediaCreatorHelper.php:31]:
```suggestion
class UriGenerator
```

File: src/Core/Infrastructure/RealTime/Api/Hypermedia/HypermediaCreatorHelper.php [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+     *   '{hostId}' => 10,
+     *   '{serviceId} => 20,
+     *    ...
+     * ]
+     *
+     * @param string $uri
+     * @param array<string, mixed> $parameters
+     * @return string
+     */
+    public function generateUri(string $uri, array $parameters): string
```
Comment (kduret) [inline src/Core/Infrastructure/RealTime/Api/Hypermedia/HypermediaCreatorHelper.php:101]:
```suggestion
    public function generateUriFromTemplate(string $uri, array $parameters): string
```

File: src/Core/Infrastructure/RealTime/Api/Hypermedia/MetaServiceHypermediaProvider.php [hunk 17]
```diff
@@ -48,7 +49,7 @@ class MetaServiceHypermediaProvider implements HypermediaProviderInterface
      */
     public function __construct(
         private ContactInterface $contact,
-        protected UrlGeneratorInterface $router
+        private HypermediaCreatorHelper $helper
```
Comment (kduret) [inline src/Core/Infrastructure/RealTime/Api/Hypermedia/MetaServiceHypermediaProvider.php:52]: update phpdoc

File: src/Core/Infrastructure/RealTime/Repository/Tag/DbReadTagRepository.php [hunk 18]
```diff
@@ -22,33 +22,71 @@
...
+        // Handle pagination
+        $request .= $this->sqlRequestTranslator->translatePaginationToSql();
+
+        $statement = $this->db->prepare($this->translateDbName($request));
+
+        foreach ($this->sqlRequestTranslator->getSearchValues() as $key => $data) {
+            $type = key($data);
+            $value = $data[$type];
+            $statement->bindValue($key, $value, $type);
+        }
```
Comment (kduret) [inline src/Core/Infrastructure/RealTime/Repository/Tag/DbReadTagRepository.php:89]: get total for pagination

**Review by kduret — COMMENTED** [review]
add unit test for `helper`

File: src/Centreon/Infrastructure/Broker/BrokerRepositoryRDB.php [hunk 19]
```diff
@@ -53,14 +54,14 @@ public function __construct(DatabaseConnection $db)
...
-                    AND ns.id = :monitoringServerId
-                WHERE config_key = :configKey
+                SELECT config_value, cfgbi.config_id AS id
+                    FROM cfg_centreonbroker_info cfgbi
+                    INNER JOIN cfg_centreonbroker AS cfgb
+                        ON cfgbi.config_id = cfgb.config_id
+                    INNER JOIN nagios_server AS ns
+                        ON cfgb.ns_nagios_server = ns.id
+                        AND ns.id = :monitoringServerId
+                    WHERE config_key = :configKey
```
Comment (kduret) [inline src/Centreon/Infrastructure/Broker/BrokerRepositoryRDB.php:64]:
```suggestion
            SELECT config_value, cfgbi.config_id AS id
                FROM cfg_centreonbroker_info cfgbi
                INNER JOIN cfg_centreonbroker AS cfgb
                    ON cfgbi.config_id = cfgb.config_id
                INNER JOIN nagios_server AS ns
                    ON cfgb.ns_nagios_server = ns.id
                    AND ns.id = :monitoringServerId
                WHERE config_key = :configKey
```

File: src/Core/Domain/RealTime/Model/Service.php [hunk 20]
```diff
@@ -633,4 +642,54 @@ public function setHasGraphData(bool $hasGraphData): self
...
+    /**
+     * @param Tag $category
+     * @return self
+     * @throws \InvalidArgumentException
+     */
+    public function addCategory(Tag $category): self
+    {
+        if (! $category instanceof Tag) {
+            throw new \InvalidArgumentException('Tag model expected.');
+        }
```
Comment (kduret) [inline src/Core/Domain/RealTime/Model/Service.php:663]: this cannot happen cause the will be a typeError

File: src/Core/Domain/RealTime/Model/Service.php [hunk 21]
```diff
@@ -187,18 +192,22 @@ public function getName(): string
...
      * @return self
+     * @throws \InvalidArgumentException
      */
-    public function addServicegroup(Servicegroup $servicegroup): self
+    public function addGroup(Servicegroup $group): self
     {
-        $this->servicegroups[] = $servicegroup;
+        if (! $group instanceof Servicegroup) {
+            throw new \InvalidArgumentException('Servicegroup model expected.');
+        }
```
Comment (kduret) [inline src/Core/Domain/RealTime/Model/Service.php:209]: this cannot happen cause the will be a typeError

Comment (kduret) [reply src/Core/Domain/RealTime/Model/Service.php:209]: this check must be in `setGroups`

File: src/Core/Domain/RealTime/Model/Service.php (hunk 20 above)
Comment (kduret) [reply src/Core/Domain/RealTime/Model/Service.php:663]: this check must be in `setCategories`

File: src/Core/Infrastructure/RealTime/Hypermedia/AbstractHypermediaProvider.php [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+     * @param string[] $topologyRoles
+     * @return boolean
+     */
+    protected function canContactAccessPages(ContactInterface $contact, array $topologyRoles): bool
+    {
+        if (! $contact->isAdmin() && ! $this->hasTopologyAccess($contact, $topologyRoles)) {
+            return false;
+        }
+
+        return true;
```
Comment (kduret) [inline src/Core/Infrastructure/RealTime/Hypermedia/AbstractHypermediaProvider.php:43]: avoid double negation
```suggestion
        return $contact->isAdmin() || $this->hasTopologyAccess($contact, $topologyRoles));
```

**Review by kduret — APPROVED** [review]

**Review by a-launois — APPROVED** [review]

**Pull request merged**